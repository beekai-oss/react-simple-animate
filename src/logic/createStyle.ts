import { Keyframes } from '../types';
import camelCaseToDash from '../utils/camelCaseToDash';

const generateKeyframes = keyframes => {
  const animationLength = keyframes.length;
  return keyframes.reduce((previous, keyframe, currentIndex) => {
    const keyframePersentage =
      animationLength === 2
        ? currentIndex * 100
        : parseFloat((100 / (animationLength - 1)).toFixed(2)) * currentIndex;

    if (typeof keyframe === 'string') {
      return `${previous} ${keyframePersentage}% {${keyframe}}`;
    }
    const keys = Object.keys(keyframe);

    if (keys.length && isNaN(+keys[0])) {
      const keyframeContent = keys.reduce(
        (acc, key) => `${acc} ${camelCaseToDash(key)}: ${keyframe[key]};`,
        '',
      );
      return `${previous} ${keyframePersentage}% {${keyframeContent}}`;
    }
    return `${previous} ${Object.keys(keyframe)[0]}% {${
      Object.values(keyframe)[0]
    }}`;
  }, '');
};

export default function createStyle({
  keyframes,
  animationName,
}: {
  keyframes: Keyframes;
  animationName: string;
}): string {
  return `@keyframes ${animationName} {${generateKeyframes(keyframes)}}`;
}
