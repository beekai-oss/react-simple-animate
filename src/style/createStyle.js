// @flow
import type { Keyframes } from '../animateKeyframes';

export default function createStyle({ keyframes, animationName }: { keyframes: Keyframes, animationName: string }) {
  const animationLength = keyframes.length;

  return `${keyframes.reduce((previous, keyframe, currentIndex) => {
    if (typeof keyframe === 'string') {
      return `${previous} ${
        animationLength === 2 ? currentIndex * 100 : parseFloat((100 / (animationLength - 1)).toFixed(2)) * currentIndex
      }% {${keyframe}}`;
    }
    // $FlowIgnoreLine:
    return `${previous} ${Object.keys(keyframe)[0]}% {${Object.values(keyframe)[0]}}`;
  }, `@keyframes ${animationName} {`)}}`;
}
