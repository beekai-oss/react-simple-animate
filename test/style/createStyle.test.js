import createStyle from '../../src/style/createStyle';

describe('createStyle', () => {
  it('should create correct style', () => {
    expect(
      createStyle({
        keyframes: ['opacity: 0;', 'opacity: 1;', 'opacity: 2;', 'opacity: 3;'],
        animationName: 'test',
        className: 'bill',
      }),
    ).toMatchSnapshot();
  });

  it('should create correct style with object', () => {
    expect(
      createStyle({
        keyframes: [{ 0: 'opacity: 0;' }, { 25: 'opacity: 1;' }, { 75: 'opacity: 2;' }, { 100: 'opacity: 3;' }],
        animationName: 'test',
        className: 'bill',
      }),
    ).toMatchSnapshot();
  });
});
