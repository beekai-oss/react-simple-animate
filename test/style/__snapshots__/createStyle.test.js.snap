// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createStyle should create correct style 1`] = `"@keyframes test { 0% {opacity: 0;} 33.33% {opacity: 1;} 66.66% {opacity: 2;} 99.99% {opacity: 3;}}"`;

exports[`createStyle should create correct style with object 1`] = `"@keyframes test { 0% {opacity: 0;} 25% {opacity: 1;} 75% {opacity: 2;} 100% {opacity: 3;}}"`;
